Automattic-Tracks-iOS/Event Logging/EventLogging.swift

```diff
@@ -0,0 +1,175 @@
+        guard exponentialBackoffTimer.delay != 0 else {
+            return nil
+        }
+
+        return exponentialBackoffTimer.nextDate
+    }
+}
+
+extension EventLogging {
+
+    /// Start uploading the next log if the queue, if possible.
+    /// This method shouldn't be called directly – it should be dispatched via `uploadNextLogFileIfNeeded`
+    private func runUploadLogs() {
+        /// Ensure that only one instance of this method is running at the same time
+        guard lock.try() else {
+            return
+        }
+
+        /// If the queue is empty, pause upload
+        guard let log = uploadQueue.first else {
+            retryUploadsAt(.distantFuture)
+            lock.unlock()
+            return
+        }
+
+        /// If the backoff timer is blocking execution, reschedule the next run
+        guard exponentialBackoffTimer.next < .now() else {
+            retryUploadsAt(exponentialBackoffTimer.next)
+            lock.unlock()
+            return
+        }
+
+        /// If the delegate is reporting that we shouldn't upload log files, pause upload to prevent an endless loop
+        guard delegate.shouldUploadLogFiles else {
+            retryUploadsAt(.distantFuture)
+            lock.unlock()
+            return
+        }
+
+        let dispatchGroup = DispatchGroup()
+        dispatchGroup.enter()
+
+        processingQueue.async {
+            dispatchGroup.enter()
+
+            do {
+                let encryptedLog = try self.encryptLog(log)
+
+                /// Upload the log
+                self.uploadManager.upload(encryptedLog) { result in
+                    if case .success = result {
+                        try? self.uploadQueue.remove(log)
+
+                        /// Reset the timer if requests are succeeding
+                        self.exponentialBackoffTimer.reset()
+                    }
+                    else {
+                        /// Wait longer between requests if they are failing
+                        self.exponentialBackoffTimer.increment()
+                        self.retryUploadsAt(self.exponentialBackoffTimer.next)
+                    }
+
+                    dispatchGroup.leave()
+                }
+            }
+            catch let err {
+                /// This is almost certainly a file error – encryption errors would assert and crash the app. This includes situations like:
+                /// - the device is out of storage space
+                /// - the file was deleted while we were reading it
+                /// - the file (or device storage system) is corrupt
+                /// Because this should be extremely rare (and is difficult to reproduce), we'll track it but it's not covered by a test case
+                CrashLogging.logError(err, userInfo: [
+                    "errorFile": #file,
+                    "errorLine": #line,
+                    "logFileUUID": log.uuid
+                ])
+            }
+
+            dispatchGroup.leave()
```

Comment:
I adjusted this in `979c827496caeb60a6140a6f5e4bc78f728f766c`. This is a tricky one to deal with, since as far as I know, this code path should be mostly impossible to hit, but in the exceedingly rare condition it might, I don't think it's worth bringing down the app for.